### PR TITLE
OpenGL-based implementation of Unicon's 2D/3D graphics

### PR DESCRIPTION
**Pull request by gkzyoung**
Signed-off-by: Gigi Young <kzyoung93@gmail.com>

### CODE REVIEW

File: src/runtime/fwindow.r [hunk 1]
```diff
@@ -1214,7 +1386,7 @@ function{1} DrawPolygon(argv[argc])
        */
       for(i = 0, j = 1; i < n; i++, j++) {
          base = warg + i * 2;
-         if (j == MAXXOBJS) {
+         if (!w->window->is_gl && j == MAXXOBJS) {
```
Comment (Jafaral) [inline src/runtime/fwindow.r:1389]: missing #ifdef GraphicsGL

File: src/runtime/Makefile [hunk 2]
```diff
@@ -2,21 +2,15 @@
 
 include ../../Makedefs
 
-
-#REPO_REV="$(shell LC_ALL=C svnversion -cn ../../ | sed -e "s/.*://" -e "s/\([0-9]*\).*/\1/" | grep "[0-9]" )"
-
-
-REPO_REV_COUNT="$(shell LC_ALL=C git rev-list --first-parent --count HEAD )"
-REPO_REV_HASH="$(shell LC_ALL=C git rev-parse --short HEAD)"
-REPO_REV="$(REPO_REV_COUNT)-$(REPO_REV_HASH)"
```
Comment (Jafaral) [inline src/runtime/Makefile:11]: Why are these lines removed ?

File: src/runtime/Makefile [hunk 3]
```diff
@@ -50,7 +44,7 @@ interp_all: update_rev
 
 # NTCONDEP is set to NTCON on windows only to add a dependency that sets NTCONSOLE
 iconx $(ICONXEXE): $(NTCONDEP) $(OBJS) ../../bin/libucommon.a ../../bin/libuconsole.a ../../bin/libgdbm.a ../../bin/libtp.a
-	$(CC) $(LDFLAGS) $(WSUBSYS) $(WSTKLDFLAG) $(RLINK) -o iconx  $(OBJS) $(XLIBS) $(RLIBS) $(XL) $(LIBSTDCPP) $(WGLIBS) ../../bin/libuconsole.a
+	$(CC) -pg $(LDFLAGS) $(WSUBSYS) $(WSTKLDFLAG) $(RLINK) -o iconx  $(OBJS) $(XLIBS) $(RLIBS) $(XL) $(LIBSTDCPP) $(WGLIBS) ../../bin/libuconsole.a
```
Comment (Jafaral) [inline src/runtime/Makefile:47]: you probably don't want to commit this.

File: src/h/graphics.h [hunk 4]
```diff
@@ -356,21 +391,34 @@ typedef struct _wdisplay {
   Display *	display;
   GC		icongc;
   Colormap	cmap;
+#ifdef GraphicsGL
+  int		nConfigs;
+  GLXFBConfig	*configs;
+  XVisualInfo   *vis;
```
Comment (Jafaral) [inline src/h/graphics.h:397]: *vis - format/whitespace compared with lines above and below

File: src/h/graphics.h [hunk 5]
```diff
@@ -742,11 +821,9 @@ typedef struct
 #define A_GLVERSION    	86
 #define A_GLVENDOR     	87
 #define A_GLRENDERER   	88
+#define A_ALPHA	 	89	
+#define A_RENDERMODE	90
```
Comment (Jafaral) [inline src/h/graphics.h:825]: format align ?

File: src/h/opengl.h [hunk 6]
```diff
@@ -12,9 +33,1215 @@
 #define U3D_QUAD_STRIP     GL_QUAD_STRIP
 #define U3D_POLYGON        GL_POLYGON
 
+#define U2D_LINE_LOOP      GL_LINE_LOOP
+#define U2D_LINE_STRIP	   GL_LINE_STRIP
+#define U2D_POLYGON	   GL_POLYGON
```
Comment (Jafaral) [inline src/h/opengl.h:38]: format

File: src/h/opengl.h [hunk 7]
```diff
@@ -12,9 +33,1215 @@
...
+#define GL2D_LINE_SOLID		0
+#define GL2D_LINE_DASHED	1
+#define GL2D_LINE_STRIPED	2
+
+/*
+ * fill styles
+ */
+#define GL2D_FILL_SOLID		0
+#define GL2D_FILL_TEXTURED	1
+#define GL2D_FILL_MASKED	2
```
Comment (Jafaral) [inline src/h/opengl.h:106]: All of the define lines  from line 50 down to here could use some format/align.

File: src/h/opengl.h [hunk 8]
```diff
@@ -12,9 +33,1215 @@
...
+#define GL2D_DRAWOP_COPY	GL_COPY	
+#define GL2D_DRAWOP_REVERSE	GL_XOR	
+
+
+/*
+ * Structures
+ */
+
+/*
+ * For a linked link color structure
```
Comment (Jafaral) [inline src/h/opengl.h:120]: link, typo ?

File: src/h/sys.h [hunk 9]
```diff
@@ -365,6 +365,11 @@
 #include <GL/glx.h>
 #endif					/* XWindows */
 #include <GL/glu.h>
+#if HAVE_LIBFREETYPE
+   #include <ft2build.h>
+   #include FT_FREETYPE_H
+   #define PNG_SKIP_SETJMP_CHECK /* to fix compile error specified by png.h */
```
Comment (Jafaral) [inline src/h/sys.h:371]: What is the  png  error ?

File: src/h/opengl.h (hunk 7 above)
Comment (gkzyoung) [reply src/h/opengl.h:106]: I took a look at my source code, and it appears that the misalignments are due to tabs being pushed by the prepended '+'

File: src/h/opengl.h (hunk 8 above)
Comment (gkzyoung) [reply src/h/opengl.h:120]: Yep, type. Thanks

File: src/h/sys.h (hunk 9 above)
Comment (gkzyoung) [reply src/h/sys.h:371]: To be honest, I don't remember. I tried compiling without it, and it runs. I'll remove it.

File: src/h/define.h [hunk 10]
```diff
@@ -0,0 +1,3 @@
+#ifndef POLL_INTERVAL
+#define POLL_INTERVAL 400
```
[the remark this answers is not in the document]
Comment (gkzyoung) [reply src/h/define.h:2]: It is a definition to remove the magic numbers from pollevent() and related functions (in rxwin.ri and rmswin.ri). Would it be okay to move it to the bottom of rmacros.h?

**Comment by Jafaral** [comment]
@gkzyoung , can you please rebase the PR on top of the master branch ?  I'm ready to merge this. Thanks!

**Comment by gkzyoung** [comment]
Just to make sure I'm understanding correctly, I should rebase my master
branch to have a single commit?

**Comment by Jafaral** [comment]
No, you can have more than one commit. I will send you info tomorrow.
